### PR TITLE
[clang] constexpr built-in reduce add function. 

### PR DESCRIPTION
**Pull request by c8ef**
Part of #51787.

This patch adds constexpr support for the built-in reduce add function. If this is the right way to go, I will add support for other reduce functions in later patches.

### CODE REVIEW

File: clang/lib/AST/ExprConstant.cpp [hunk 1]
```diff
@@ -13528,6 +13528,20 @@ bool IntExprEvaluator::VisitBuiltinCallExpr(const CallExpr *E,
     return Success(DidOverflow, E);
   }
 
+  case Builtin::BI__builtin_reduce_add: {
+    APValue Source;
+    if (!EvaluateAsRValue(Info, E->getArg(0), Source))
+      return false;
+
+    auto SourceLen = Source.getVectorLength();
```
Comment (RKSimon) [inline clang/lib/AST/ExprConstant.cpp:13536]: (style) Avoid auto

**Review by RKSimon — COMMENTED** [review]
This seems reasonable to me - I'd recommend you take a look at the min/max reductions to see if signed/unsigned can be correctly handled with this approach.

File: clang/test/Sema/constant_builtins_vector.cpp [hunk 2]
```diff
@@ -723,3 +723,9 @@ not within the bounds of the input vectors; index of -1 found at position 0 is n
 permitted in a constexpr context}}
         vector4charConst1,
         vector4charConst2, -1, -1, -1, -1);
+
+static_assert(__builtin_reduce_add((vector4char){}) == 0);
```
Comment (RKSimon) [inline clang/test/Sema/constant_builtins_vector.cpp:727]: How is this handled? I'd have expected `Source.getVectorElt(0).getInt()` to have failed?

File: clang/lib/AST/ExprConstant.cpp [hunk 3]
```diff
@@ -13528,6 +13528,24 @@ bool IntExprEvaluator::VisitBuiltinCallExpr(const CallExpr *E,
...
+      return false;
+
+    unsigned SourceLen = Source.getVectorLength();
+    APSInt Reduced = Source.getVectorElt(0).getInt();
+    for (unsigned EltNum = 1; EltNum < SourceLen; ++EltNum) {
+      if (!CheckedIntArithmetic(
+              Info, E, Reduced, Source.getVectorElt(EltNum).getInt(),
+              Reduced.getBitWidth() + 1, std::plus<APSInt>(), Reduced)) {
+        return false;
+      }
```
Comment (Fznamznon) [inline clang/lib/AST/ExprConstant.cpp:13543]: I think you can elide the braces here
```suggestion
              Reduced.getBitWidth() + 1, std::plus<APSInt>(), Reduced))
        return false;
```

**Review by Fznamznon — APPROVED** [review]
LGTM, thanks!

**Comment by RKSimon** [comment]
Thanks @c8ef - are you OK to add the remaining integer reduction cases next? Also, please can you also update the LanguageExtensions docs page to mention that they are constexpr legal?

**Pull request merged**